store/src/java/com/zimbra/cs/event/logger/EventLogger.java

```diff
@@ -197,29 +210,32 @@ private EventNotifier(Map<String, EventLogHandler.Factory> knownFactories, Map<S
         @Override
         public void run() {
             try {
-                while (!Thread.currentThread().isInterrupted()) {
-                    consume(eventQueue);
+                Event event = null;
+                while (!shutdownExecutor.get()) { //don't process any more events if shutdownExecutor is set to true.
+                    /* poll is used so we can check if shutdownExecutor is set every 3 seconds.
+                    if take was used the code would block on it in case of empty event queue. */
+                    event = eventQueue.poll(POLL_TIMEOUT, TimeUnit.SECONDS);
+                    if (event != null && event != POISON_PILL) {
+                         notifyEventLogHandlers(event);
+                    }
+                }
+                //drain the event queue till POISON_PILL is encountered.
+                if (drainQueueBeforeShutdown.get() && !(event != null && event != POISON_PILL)) {
+                    ZimbraLog.event.debug("Draining the queue");
+                    event = eventQueue.poll(); //wait till event is added to the queue. In case of empty queue wait for POISON_PILL
+                    while (event != null && event != POISON_PILL) { //exit loop if POISON_PILL is encountered
```

Comment:
Would processing all messages after the POISON_PILL is encountered be such a bad thing?
As soon as the queue empties then it could exit.  This would ensure that nothing is 'lost' even if produced after the poisoning.   I presume the events are not produced at a faster rate than they can be processed.

Reply:
@tmclane Actually, thats a good point. In that case we would not need any poison pills. In the drainQueueBeforeShutdown if block, if we get a null back from a poll it will exit. I will test this out. Thanks for pointing this out.